qcodes/dataset/data_set_cache.py

```diff
@@ -0,0 +1,99 @@
+from typing import TYPE_CHECKING, Dict, Optional
+
+import numpy as np
+
+from qcodes.dataset.sqlite.queries import (
+    get_interdeps_from_result_table_name, completed, get_non_dependencies,
+    get_parameter_data_for_one_paramtree)
+
+if TYPE_CHECKING:
+    import pandas as pd
+    from .data_set import DataSet, ParameterData
+
+
+class DataSetCache:
+    """
+    The DataSetCache contains a in memory representation of the
+    data in this dataset as well a a method to progressively read data
+    from the db as it is written. The cache is available in the same formats
+    as :py:class:`.DataSet.get_parameter_data` and :py:class:`.DataSet.get_data_as_pandas_dataframe`
+
+    """
+
+    def __init__(self, dataset: 'DataSet'):
+        self._dataset = dataset
+        self._data: ParameterData = {}
+        self._read_status: Dict[str, int] = {}
+        self._loaded_from_completed_ds = False
+
+    def load_data_from_db(self) -> None:
+        """
+        Loads data from the dataset into the cache:
+        If the dataset is marked completed and data has already been loaded
+        no load will be performed.
+        """
+        if self._loaded_from_completed_ds:
+            return
+        self._dataset._completed = completed(self._dataset.conn, self._dataset.run_id)
+        if self._dataset.completed:
+            self._loaded_from_completed_ds = True
+
+        parameters = get_non_dependencies(self._dataset.conn, self._dataset.table_name)
```

Comment:
`get_non_dependencies` seems to be used only in this file and is inside calling `get_interdeps_from_result_table_name`. i'd suggest to get rid of `get_non_dependencies` and instead of it's call here execute `parameters = tuple(ps.name for ps in interdeps.non_dependencies)` on the `interdeps` returned by `get_interdeps_from_result_table_name`

Reply:
Sure we can do this. As far as I remember this is not new code. I simply extracted part of a function into its own